src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs

```diff
@@ -225,6 +230,27 @@ public string GetInstalledPackageFilePath(PackageIdentity packageIdentity)
         }
 
         /// <summary>
+        /// Get the path to the package nuspec.
+        /// </summary>
+        public string GetInstalledManifestFilePath(PackageIdentity packageIdentity)
+        {
+            // Check the expected location before searching all directories
+            var packageDirectory = PackagePathResolver.GetInstallPath(packageIdentity);
+            var manifestName = PackagePathResolver.GetManifestFileName(packageIdentity);
+
+            var installPath = Path.GetFullPath(Path.Combine(packageDirectory, manifestName));
+
+            // Keep the previous optimization of just going by the existance of the file if we find it.
+            if (File.Exists(installPath))
+            {
+                return installPath;
+            }
+
+            // Don't look in non-normalized paths for nuspec
```

Comment:
-ExcludeVersion isn't valid for nuget restore. If you restore into a PackagesDirectory with a package that has been installed with -ExcludeVersion, nuget restore will ignore that package and install a new package with a normalized path. Also -ExcludeVersion would create "newtonsoft.json\newtonsoft.json.nupkg" which wouldn't be matched by PackageExistsInPackagesFolder either as it expects the version to either be in the directory or the filename.
Any other scenarios you can think of?